### PR TITLE
perf(checkpointing): describe prepended-axis shards with the checkpointable path

### LINKED ISSUE #7275: [ENHANCEMENT] Reduce torch_dist checkpoint preparation overhead for prepended-axis tensors
**Problem**

We observed substantial CPU overhead during distributed checkpoint saving in a v0.16.0-based fork. A large part of the training pause was spent in `mcore_to_pyt_state_dict`, converting shard metadata before asynchronous checkpoint writing could begin.

The affected case is a regular, non-flattened `ShardedTensor` with `prepend_axis_num > 0`, commonly used to represent layer/expert axes. These tensors are routed through the legacy PyTorch `ShardedTensor` conversion even when the PyTorch version supports the checkpointable path.

The legacy conversion enumerates the entire global shard grid, including remote shards, for each converted key. Consequently, preparation cost grows with the number of fragments across the logical axes, rather than only the shards held locally. This becomes expensive for MoE checkpoints with many layer/expert shards.

Because this conversion happens synchronously before the async write request is returned, enabling asynchronous checkpointing does not hide this overhead. The same routing condition remains in the [inspected upstream revision](https://github.com/NVIDIA/Megatron-LM/blob/f6c33bde4cdc10805ee7157db746d47b758f5f02/megatron/core/dist_checkpointing/strategies/torch.py#L293-L315).

**Observed impact**

Our workload was a xxB MoE model with xx layers, xx experts and EP=x, running on eight GPUs with PyTorch 2.10.0. The profiling record identified 14,832 of 15,447 sharded-tensor records on the affected path.

With the downstream workaround, recorded `mcore_to_pyt_state_dict` time dropped from **32.1 seconds to 0.13 seconds**. This measures conversion only, not total checkpoint completion time. The timings are from our downstream fork; upstream was checked by source inspection rather than a fresh benchmark.

**Downstream workaround**

Our MR allows eligible prepended-axis tensors to use the existing checkpointable path. A prepended axis has extent one within each local chunk, so we expose those singleton dimensions consistently in the local chunk metadata and tensor view. The adapter can then describe local shards directly, avoiding enumeration of the global grid.

This preserves the logical tensor shapes and offsets and uses views rather than copying dense tensor payloads. Existing version and flattened-shard restrictions are retained. Quantized tensors need representation-preserving views; GPU validation of the final FP8 changes is still pending.

Would maintainers consider extending the checkpointable path to this case? We would appreciate feedback on this approach and any compatibility requirements that should be addressed before upstreaming it.

### PR DESCRIPTION
**Pull request by LiRunGuo**
## Motivation

`mcore_to_pyt_state_dict` routes every `ShardedTensor` with a non-zero `prepend_axis_num`
through the legacy PyTorch `ShardedTensor` conversion
(`sharded_tensor_to_torch_sharded_tensor`). That conversion enumerates the *entire global
shard grid* — remote shards included — for each converted key:

```python
for fragment_offsets in product(*map(range, some_sh_ten.axis_fragmentations)):
```

so its cost grows with the number of fragments across all logical axes, not with the number of
locally held shards.

Prepended axes are not an exotic corner case (closes #7275):

* `TransformerBlock.sharded_state_dict` passes `(0, global_layer_offset, num_layers)` as the
  `sharded_offsets` of every layer whenever the stack is homogeneous, and
  `sharded_state_dict_default` turns those offsets into a **prepended layer axis** of every
  parameter of the block.
* `SequentialMLP.sharded_state_dict` adds a **prepended global-expert axis** on top of it for
  each routed expert.

For a 48-layer MoE model with 128 experts and EP=8 the per-rank local shards for one weight key
therefore describe a 48 x 128 grid — 4.7M grid entries per rank in total — and all of that work
happens **synchronously before the async write request is returned**, so asynchronous
checkpointing cannot hide it from the training loop.

## What changed

A prepended axis has extent **one** within a local chunk (that is what `prepend_axis_num` means:
the local data covers a single index along the prepended axes). The DCP checkpointable path can
describe that directly, so prepended-axis shards no longer need the legacy conversion:

* `CheckpointableShardedTensor` now reports prepended axes as singleton dimensions of the local
  chunk (`sizes = (1,) * prepend_axis_num + local_shape`) and exposes the local data through a
  matching view (`data.view((1,) * prepend_axis_num + local_shape)`) — a view, no payload copy.
* `mcore_to_pyt_state_dict` routes prepended-axis shards to the checkpointable path, keeping the
  existing restrictions (PyTorch >= 2.6, regular grid, no flattened shards).

The resulting DCP metadata is **identical** to the metadata produced by the legacy conversion
(same chunk offsets, chunk sizes, global size and dtype — asserted by the new tests), so:

* checkpoints written before this change still load, and
* checkpoints written after it are indistinguishable from the old format.

## Measurements (8xH200, PyTorch 2.10, `torch_dist`)

Local (non-TransformerEngine) MoE model, 48 layers, 128 experts, EP=8, 1,109 local shards per
rank, 1,646 unique keys, 4,734,725 grid entries per rank:

| phase | before | after |
| --- | --- | --- |
| `mcore_to_pyt_state_dict` | 11.71 s | **0.07 s** (165x) |
| synchronous prefix of `async_save()` | 17.75 s | **5.22 s** (3.4x) |
| `save_preprocess` (unchanged) | 0.58 s | 0.60 s |

A synthetic prepended-axis workload with 3,072 keys/rank x 128 grid entries (393,216 entries)
lands in the same place: 3.69 s -> 0.027 s.

## Testing

* New: `tests/unit_tests/dist_checkpointing/test_checkpointable.py::TestPrependedAxisCheckpointable`
  — write-item metadata parity with the legacy conversion, save/load round trip, and loading a
  checkpoint written by the legacy conversion.
* New: `tests/unit_tests/dist_checkpointing/models/test_moe_expert_sharding.py` — MoE model
  round trips for EP=1/2/4 and EP=4 -> EP=2 resharding (asserts the prepended layer/expert axes
  are present). Runs without TransformerEngine.
* `pytest tests/unit_tests/dist_checkpointing/{test_checkpointable,test_torch_dist,
  test_serialization,test_validation,test_local,test_fully_parallel,test_replication}.py`
  on 8 GPUs: 65 passed, 2 skipped.
* `black`, `isort`, `ruff`, `pylint` clean on the changed files.

## Follow-ups (not in this PR)

The legacy branch of `mcore_to_pyt_state_dict` (and `sharded_tensor_to_torch_sharded_tensor`)
is now reachable only for `torch < 2.6`; removing it once the minimum PyTorch version allows
would be a separate cleanup.

### CODE REVIEW

**Review by dimapihtar — APPROVED** [review]

**Comment by Connor-XY** [comment]
@NVIDIA/core-adlr @NVIDIA/core-nemo The checkpointing expert review is approved on the current head (`471fa1c`), and there are no outstanding inline review threads. Could the already-requested final-review teams complete the repository-level review? This is the prepended-axis checkpoint conversion optimization tracked in #7275.

**Review by Connor-XY — APPROVED** [review]
